lib/dao.js

```diff
@@ -586,6 +586,195 @@ DataAccessObject.updateOrCreate = DataAccessObject.upsert = function upsert(data
 };
 
 /**
+ * Replace or insert a model instance: replace exiting record if one is found, such that parameter `data.id` matches `id` of model instance;
+ * otherwise, insert a new record.
+ *
+ * NOTE: No setters, validations, or hooks are applied when using upsert.
+ * @param {Object} data The model instance data
+ * @param {Object} [options] Options for replaceOrCreate
+ * @param {Function} cb The callback function (optional).
+ */
+
+DataAccessObject.replaceOrCreate  = function replaceOrCreate(data, options, cb) {
+  var connectionPromise = stillConnecting(this.getDataSource(), this, arguments);
+  if (connectionPromise) {
+    return connectionPromise;
+  }
+
+  if (options === undefined && cb === undefined) {
+    if (typeof data === 'function') {
+      cb = data;
+      data = {};
+    }
+  } else if (cb === undefined) {
+    if (typeof options === 'function') {
+      cb = options;
+      options = {};
+    }
+  }
+
+  cb = cb || utils.createPromiseCallback();
+  data = data || {};
+  options = options || {};
+
+  assert(typeof data === 'object', 'The data argument must be an object');
+  assert(typeof options === 'object', 'The options argument must be an object');
+  assert(typeof cb === 'function', 'The cb argument must be a function');
+
+  var hookState = {};
+
+  var self = this;
+  var Model = this;
+  var connector = Model.getConnector();
+
+  var id = getIdValue(this, data);
+  if (id === undefined || id === null) {
+    return this.create(data, options, cb);
+  }
+
+  var context = {
+    Model: Model,
+    query: byIdQuery(Model, id),
+    hookState: hookState,
+    options: options
+  };
+  Model.notifyObserversOf('access', context, doReplaceOrCreate);
+
+  function doReplaceOrCreate(err, ctx) {
+    if (err) return cb(err);
+
+    var isOriginalQuery = isWhereByGivenId(Model, ctx.query.where, id);
+    if (connector.replaceOrCreate && isOriginalQuery) {
+      var context = {
+        Model: Model,
+        where: ctx.query.where,
+        data: data,
+        hookState: hookState,
+        options: options
+      };
+      Model.notifyObserversOf('before save', context, function(err, ctx) {
+        if (err) return cb(err);
+
+        data = ctx.data;
+        var update = data;
+        var inst = data;
+        if (!(data instanceof Model)) {
+          inst = new Model(data, { applyDefaultValues: false });
+        }
+        update = inst.toObject(false);
```

Comment:
I think this whole block (L658-L664) can be removed, because the hook is manipulating directly the instance, not `data`. Please check the implementation of `create`.